File: libandroidapi/src/main/java/systems/altimit/libandroidapi/modules/AndroidFS.java
/*
 * Copyright (c) 2018 Altimit Community Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or imp
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package systems.altimit.libandroidapi.modules;

import android.webkit.JavascriptInterface;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.OutputStreamWriter;

/**
 * Created by felixjones on 24/02/2018.
 */
public class AndroidFS {

    public static final String INTERFACE_NAME = "__android_api_fs";

    @JavascriptInterface
    public String fsExistsSync(String path) {
        return new File(path).exists() ? "true" : "false";
    }

    @JavascriptInterface
    public void fsMkdirSync(String path) {
        new File(path).mkdirs();
    }

    @JavascriptInterface
    public void fsWriteFileSync(String file, String data) {
        OutputStreamWriter osw = null;
        try {
            osw = new OutputStreamWriter(new FileOutputStream(new File(file)), "UTF-8");
            osw.write(data);
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (osw != null) {
                    osw.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    @JavascriptInterface
    public String fsReadFileSync(String path) {
        StringBuilder text = new StringBuilder();
        BufferedReader br = null;
        try {
            br = new BufferedReader(new FileReader(new File(path)));
            String line;
            while ((line = br.readLine()) != null) {
                text.append(line);
                text.append('\n');
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (br != null) {
                    br.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return text.toString();
    }

    @JavascriptInterface
    public void fsUnlinkSync(String path) {
        new File(path).delete();
    }

}

File: libandroidapi/src/main/res/raw/android_fs.js
/*
 * Copyright (c) 2018 Altimit Community Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or imp
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

( function() {

    /**
     * FileCache
     */
    function FileCache() {
        throw new Error( 'This is a static class' );
    }
    ( function() {

        FileCache.limit = 1024 * 1024; // ~1 MiB
        FileCache._items = {};

        FileCache.add = function( key, value ) {
            this._items[key] = {
                contents: value,
                touch: Date.now(),
                key: key
            };
            this._truncateCache();
        };

        FileCache.get = function( key ) {
            if ( this._items[key] ) {
                var item = this._items[key];
                item.touch = Date.now();
                return item.contents;
            }
            return null;
        };

        FileCache.remove = function( key ) {
            delete this._items[key];
        };

        FileCache._truncateCache = function(){
            var items = this._items;
            var sizeLeft = FileCache.limit;

            Object.keys( items ).map( function( key ) {
                return items[key];
            } ).sort( function( a, b ) {
                return b.touch - a.touch;
            } ).forEach( function( item ) {
                if ( sizeLeft > 0 ){
                    var contents = item.contents;
                    sizeLeft -= contents.length;
                } else {
                    delete items[item.key];
                }
            }.bind( this ) );
        };

    } )();

    var _exists = {};

    androidRequire.fs = {

        existsSync: function( path ) {
            if ( !_exists[path] ) {
                _exists[path] = {
                    value: __android_api_fs.fsExistsSync( path ) == 'true'
                }
            }
            return _exists[path].value;
        },

        mkdirSync: function( path, mode ) {
            return __android_api_fs.fsMkdirSync( path );
        },

        writeFileSync: function( file, data, options ) {
            _exists[file] = {
                value: true
            };

            FileCache.add( file, data );
            return __android_api_fs.fsWriteFileSync( file, data );
        },

        readFileSync: function( path, options ) {
            var contents = FileCache.get( path );
            if ( !contents ) {
                contents = __android_api_fs.fsReadFileSync( path );
                if ( !!contents ) {
                    FileCache.add( path, contents );
                }
            }
            return contents;
        },

        unlinkSync: function( path ) {
            _exists[path] = {
                value: false
            };

            FileCache.remove( path );
            return __android_api_fs.fsUnlinkSync( path );
        },

    };

} )();